src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,185 @@
 #include <boost/wave/cpplexer/cpp_lex_token.hpp>
 #include <boost/wave/cpplexer/cpp_lex_iterator.hpp>
 
-#define TCPP_IMPLEMENTATION
-#include <tcpp/source/tcppLibrary.hpp>
-#undef TCPP_IMPLEMENTATION
+//#define TCPP_IMPLEMENTATION
+//#include <tcpp/source/tcppLibrary.hpp>
+//#undef TCPP_IMPLEMENTATION
 
 using namespace nbl;
 using namespace nbl::asset;
 using Microsoft::WRL::ComPtr;
 
 static constexpr const wchar_t* SHADER_MODEL_PROFILE = L"XX_6_7";
 
+
+
 namespace nbl::asset::hlsl::impl
 {
     struct DXC {
         ComPtr<IDxcUtils> m_dxcUtils;
         ComPtr<IDxcCompiler3> m_dxcCompiler;
     };
+
+
+    // for including builtins 
+    struct load_file_or_builtin_to_string
+    {
+        template <typename IterContextT>
+        class inner
+        {
+        public:
+            template <typename PositionT>
+            static void init_iterators(IterContextT& iter_ctx,
+                PositionT const& act_pos, boost::wave::language_support language)
+            {
+                typedef typename IterContextT::iterator_type iterator_type;
+
+                std::string s(iter_ctx.filename.begin(), iter_ctx.filename.end());
+                auto icf = iter_ctx.ctx.get_hooks().preprocessOptions.includeFinder; // TODO it's null?
+                //std::optional<std::string> st = icf.getIncludeStandard(p, s);
+                //unfortunately have to call get_resource_runtime here again
+                auto builtin_pair = nbl::builtin::get_resource_runtime(s);
+                if (builtin_pair.first) //not a nullptr
+                {
+                    iter_ctx.instring = std::string(builtin_pair.first, builtin_pair.first+builtin_pair.second);
+                }
+                else {
+                    //TODO use nbl filesystem
+                    // read in the file
+                    boost::filesystem::ifstream instream(iter_ctx.filename.c_str());
+                    if (!instream.is_open()) {
+                        BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                            bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                        return;
+                    }
+                    instream.unsetf(std::ios::skipws);
+
+                    iter_ctx.instring.assign(
+                        std::istreambuf_iterator<char>(instream.rdbuf()),
+                        std::istreambuf_iterator<char>());
+                }
+                iter_ctx.first = iterator_type(
+                    iter_ctx.instring.begin(), iter_ctx.instring.end(),
+                    PositionT(iter_ctx.filename), language);
+                iter_ctx.last = iterator_type();
+            }
+
+        private:
+            std::string instring; //todo change to IReferenceCounted file
```

Comment:
actually thats fine, because you'll need to add `#line` directives to the beginning and end of an included file

This is for DXC to know what file each "block" of source came from

prelude and epilogue basically.

Reply:
from what i have seen, boost already adds those, albeit gets confused with builtins and gives a wrong abs path

it produces something like 
`.../Nabla/examples_tests/02_HelloCompute/bin/nbl/builtin/hlsl/shapes/circle.hlsl`
where as the correct path should be 
`.../Nabla/include/nbl/builtin/hlsl/shapes/circle.hlsl`